rasa/engine/validation.py

```diff
@@ -0,0 +1,324 @@
+import inspect
+import logging
+import typing
+from typing import Optional, Callable, Text, Tuple, Dict, Type, Any, Set
+
+import dataclasses
+import typing_utils
+
+from rasa.engine.exceptions import GraphSchemaValidationException
+from rasa.engine.graph import (
+    GraphSchema,
+    GraphComponent,
+    SchemaNode,
+    Fingerprintable,
+    ExecutionContext,
+)
+from rasa.engine.storage.resource import Resource
+from rasa.engine.storage.storage import ModelStorage
+from rasa.nlu import components
+
+
+@dataclasses.dataclass
+class ParameterInfo:
+    """Stores metadata about a function parameter."""
+
+    type_annotation: Any
+    is_kwargs: bool
+    has_default: bool
+
+
+_keywords_expected_types: Dict[Text, Any] = {
+    "resource": Resource,
+    "execution_context": ExecutionContext,
+    "model_storage": ModelStorage,
+    "config": Dict[Text, Any],
+}
+
+
+def validate(
+    schema: GraphSchema, language: Optional[Text], is_train_graph: bool
+) -> None:
+    """Validates a graph schema.
+
+    This checks that the graph structure is correct (e.g. all nodes pass the correct
+    things into each other) and validates the single graph components.
+
+    Args:
+        schema: The schema which needs validating.
+        language: Used to validate if all components support the language the assistant
+            is used in. If the language is `None`, all components are assumed to be
+            compatible.
+        is_train_graph: Whether the graph is used for training.
+
+    Raises:
+        GraphSchemaValidationException: If the validation failed.
```

Comment:
Very Optional: Did you consider collecting all the errors and returning/displaying them all at the end rather than the user having to do one-by-one?

Reply:
mhm, interesting idea 🤔 I haven't considered that. Might also obfuscate the code a lot as some validations require certain information to be present from earlier validations (e.g. checking type hints requires a successful retrieval of the function and its type hints in the first place). I'd go for this (simpler) solution for now as most graphs should be generated from the `Recipe`s anyway.